### ISSUE TITLE
Procedural parts support

### ISSUE THREAD

**Issue by jwvanderbeck** [jwvanderbeck/TestFlight#44]
Should be able to tie in and allow different settings per size or radius or something.

**Comment by jwvanderbeck** [comment]
Most of the code is in place with the Interop system.  Waiting on a release of this then can look into getting it into TweakScale

**Comment by jwvanderbeck** [comment]
Interop system is in place.  Need to see about getting this added to Procedural Parts.

**Comment by jwvanderbeck** [comment]
Still waiting on the ProcParts project to add support

**Comment by jwvanderbeck** [comment]
Dependent on: https://github.com/Swamp-Ig/ProceduralParts/issues/109

**Comment by jwvanderbeck** [comment]
@NathanKell Did this ever happen?

**Comment by NathanKell** [comment]
Uh. I'm not sure, it's been a while. I think I did, but 1.0.5 blurred all memories. >.>

**Comment by NathanKell** [comment]
If not I'll do it now.

**Comment by anxcon** [comment]
@NathanKell @jwvanderbeck 
oh look this issue exists :) guess i should poke this beehive

uh by reading above TF has al the stuff needed done or ? 
NathanKell said the interop stuff was added/done in PParts :)

P.S. testflight is confusing code in this area

**Comment by jwvanderbeck** [comment]
The interop on th TestFlight side was done ages ago it was just waiting on pparts to work with it

Sent from Outlook Mobile




On Wed, Feb 24, 2016 at 2:24 PM -0800, "anxcon" <notifications@github.com> wrote:












@NathanKell @jwvanderbeck 

oh look this issue exists :) guess i should poke this beehive



uh by reading above TF has al the stuff needed done or ? 

NathanKell said the interop stuff was added/done in PParts :)



P.S. testflight is confusing code in this area

**Comment by anxcon** [comment]
so....everything is done? nothing is needed on TF side?
assuming so, what to do on TF configs etc to enable it all to work?:P

**Comment by NathanKell** [comment]
Ah, then I'm confused. What *should* I have added to proc parts then? I wasn't aware you had a spec for the interops you wanted.

**Comment by NathanKell** [comment]
Ah, you had suggestions at the top of the issue on the PP repo. Which I missed. Are those what's coded?

**Comment by jwvanderbeck** [comment]
> Ah, you had suggestions at the top of the issue on the PP repo. Which I missed. Are those what's coded?

Well nothing is coded.  That is actually the idea behind the interop is that TF has absolutely no innate knowledge of any other plugin or what it is doing.  It simply exposes the registered values to configs, allowing you to use them for differentiating a part.  So whatever you added to PParts is what is available to config files for use in calling out a part for different TF stats.

For reference: https://github.com/KSP-RO/TestFlight/wiki/Interop-System

**Comment by NathanKell** [comment]
Ok, then I need to recode what I wrote, clearly. :]

**Comment by NathanKell** [comment]
Err, let me quickly ask then. If I am passing multiple values, how do I constrain TF configs to apply to the intersection? The interop I am familiar with (RF: tank volume, or engine config) is one-per-part at a time.

**Comment by jwvanderbeck** [comment]
Trying to refresh my memory and my wiki is kind of contradicting itself, so don't quote me 100% on this yet, but pretty sure you need to use the [Advanced Configuration Syntax](https://github.com/KSP-RO/TestFlight/wiki/Advanced-Configuration-Syntax).  However that wiki page seems to make references to PParts identifiers that never existed so that is confusing me a bit.  I think the syntax is correct but just replace the indicated PParts identifiers with the actual ones you are registering in the Interop.

**Comment by jwvanderbeck** [comment]
I'm confused because at some point in time I wrote this Wiki page on working with PParts: https://github.com/KSP-RO/TestFlight/wiki/Procedural-Parts-Support

But why did I write that if said interops weren't in place? I need to dig a bit more.

**Comment by NathanKell** [comment]
Ok, dang. I was hoping TF could handle this stuff internally. The idea was that all tanks of <=1m diameter and <=1m length  would be in one pool and share flight data, and above that there would be one pool each for each combination of
1. is it cone or fillet?
2. length rounded to nearest 25cm
3. diam1 rounded to nearest 25cm
4. diam2 rounded to nearest 25cm.

I *really* don't want to have hundreds of TF modules on a single part, each listening for a different possible configuration.

**Comment by NathanKell** [comment]
err, treating cylinders and bezier cones as cones.

**Comment by jwvanderbeck** [comment]
How would you handle it *inside* TF? The point behind the interop was to give that power to the configs and therefore not have mod specific code inside TF which is just nasty :)

**Comment by anxcon** [comment]
just thinking, have a single spot in TF (or PParts? idk ) to "encode" the variables to a single configuration name, and it runs off that, ie turn L D1 D1 into hex etc
0-1=0, 1-1.25=1.....2.75-3.0=8.......3.75-4.0=C......leaves 9.5-9.75=Z
maybe 9.5+ is Z, or move some stuff around so >5.0 is in 0.5 steps etc
lets there be a single 3 char tag, prefix modname etc
(just not sure where id put this idea XD )

also on a side note
https://github.com/KSP-RO/TestFlight/wiki/Interop-System
examples are calling for TestFlightCore, then invoke method AddInteropValue, but that method is in the module TestFlightInterop, not seeing it in core unless i'm blind which is happening latwely, not quite sure where TF runs around with it after that though

**Comment by anxcon** [comment]
accually....on that point...could skip the interop bit, just encode that tag (in PParts), and simply set to configuration field of TestFlightCore via reflection

il shutup now :)

**Comment by jwvanderbeck** [comment]
But a config can do the same thing.  It can interrogate multiple properties and ranges of values in a single configuration line.

As for TestFlightCore - that is just a static front facing interface to make things easier for mods.

**Comment by anxcon** [comment]
@jwvanderbeck 
uh suggestions for PP config then?:3 to do the above in 0.25m steps

**Comment by jwvanderbeck** [comment]
> uh suggestions for PP config then?:3 to do the above in 0.25m steps
My brain is beyond dead after a bad day at work, but if you could show me an example config in "english" I can probably show you how it would look in syntax

**Comment by anxcon** [comment]
https://github.com/Swamp-Ig/ProceduralParts/blob/master/Parts/Tanks/2TankRCS.cfg is a typical/basic PParts tank for stock monopropellant, no idea what TF config would be for such

**Comment by jwvanderbeck** [comment]
I think the problem here is the combinations.  You can write a config that says essentially "Is betwen 1 and 1.25m in length, diam1 and diam2" easily enough but you have 3 properties with X ranges and therefore Y different combinations of possabilities.

L 1-1.25, D1 1-1.25, D2 1-1.25
L 1-1.25, D1 1-1.25, D2 1.25-1.5
L 1-1.25, D1 1-1.25, D2 1.5-1.75
L 1-1.25, D1 1-1.25, D2 1.75-2
L 1-1.25, D1 1.25-1.50, D2 1-1.25
L 1-1.25, D1 1.50-1.75, D2 1-1.25
L 1-1.25, D1 1.75-2, D2 1-1.25

etc, etc.  IT grows very rapidly :(

Hmmmmmmmm

**Comment by NathanKell** [comment]
Yep, that's the problem. And it's 4, because fillet tanks need to be handled differently from cone/cyl (although that only doubles the total, it doesn't multiply it by 400^3)

**Comment by anxcon** [comment]
could drop count some, diam 1.0, 1.25, 1.5....3.0, 3.5, 4.0, 5.0 is 13, anything bigger have a 14th, length 1, 2, 3, 4, 5, 7, 10, inf would be 8, skip shape, 8*13=104 still a bit but.....

**Comment by jwvanderbeck** [comment]
Seems like a player problem too, a design problem I mean.  Let's say you do set up all those configs.  That still means you are asking the player to manage hundreds of part reliabilities which doesn't seem like a wise design.

**Comment by anxcon** [comment]
stock tanks is seperate per tank, nothing additional is asked, though some inherit would help a fair bit

**Comment by NathanKell** [comment]
On the contrary, it's an excellent decision! It encourages the reuse of stage designs! That's the whole point. :P
(The real ideal is to have diameter changes lead to a new data pool, i.e. new tooling, but length changes have tech transfer, to represent stage stretches)

**Comment by anxcon** [comment]
agree ^

though that is more RO thinking, those silly people who think stock is a game would prolly prefer 1 size fits all >.> 

so with the above goal.....would it work for PParts to encode a line > reflection it to configuration? seems like itd get the new data pool anyways

**Comment by jwvanderbeck** [comment]
> It encourages the reuse of stage designs!

No I do get that, but it feels like a round-about way of going at it I guess.

**Comment by jwvanderbeck** [comment]
Two things strike me the most...

1. Is every 25cm really necessary?
2. Isn't the trouble really more about diameter than length? 

It would seem to me you could achieve the same end goal with far less configs if you allow for more ranges of length to be considered the same part, and maybe go with something a bit more lenient on diameter like 50cm.  If you did that and allowed for larger ranges of *length* I think you could cut it down to not that many configs, but still enough to discourage major changes in tank design.

**Comment by NathanKell** [comment]
We're talking 1m diameter to 20m diameter. *Times* however many length cases, *times* 1m diameter to 20m diameter again (for cone/fillet/whatever secondary diameter), and then doubled. That's a quite insane number of configs. Also, am I wrong in interpreting things, or does every config require a set of modules? I don't even want to think about how many components that is.

50cm is, well, let's look at the bottom end. That means a 1.5m diameter stage (Agena) would be the same as a 2m diameter stage (Block L IIRC? Or pick your 6.5ft diameter stage). In hard terms, that means you could nearly double your propellant load (1.8x) for a given length...without it counting as a new stage. Let alone going from Able to AbleStar without requiring new testing.

The thing is, IMO it should be conceptualized as unlocking a new part. Tech transfer, sure, but...it's still a new part, effectively.

**Comment by jwvanderbeck** [comment]
On the one hand we have configs.  It is technically doable, and is cleaner in that you aren't hardcoding PParts handling into Test Flight.

On the other hand you have hardcoding support into Test Flight, which I strongly oppose, but it makes things easier on the config side.

How about a middle ground.  TF is after all designed to be modular.  We're almost exclusively talking about fuel tanks here so maybe the middle ground is to write up a custom pparts specific modules to plug into TF.

**Comment by jwvanderbeck** [comment]
> 50cm is, well, let's look at the bottom end. That means a 1.5m diameter stage (Agena) would be the same as a 2m diameter stage (Block L IIRC? Or pick your 6.5ft diameter stage).

Actually in my mind those would be separate. 1m, 1.5m, 2m, 2.5m etc would each be a separate part.  That said I had no idea we were talking all the way up to 20m.  That is..big :)

**Comment by NathanKell** [comment]
N-1 lower diameter is 15m IIRC. God only knows what Eotena was @ferram4
Yeah, an extra module makes sense here I think. With KSPAssembly it can seamless load (or fail to load) in the presence or absence of PP respectively and therefore avoid reflection.

**Comment by NathanKell** [comment]
and 1m, 1.5m...that means that the numbers just get shifted so we talk about Able (.8m/2.5ft) vs something 1.24m (4ft) in diameter.

**Comment by anxcon** [comment]
while i get why shape should probably effect it, what about just going by volume? far far less configs could in theory be made, but then a long skinny tank is the same as a pancake tank....considering it could save alot of configs though, perhaps still ahead having length/width ratio value to counter the issue - im far from against limiting, id except new data for exact sizing ;p but figured id throw out the idea, PParts already has volume, and L/W ratio would be simple to calc

**Comment by jwvanderbeck** [comment]
I need a little time to dig through and digest old code (that is always fun) but I am pretty sure this is doable with custom plugins to TF, because the plugins themselves are responsible for determining if they apply to a part or not.  In other words, when you use any given `configuration` string to attach a module to a specific part configuration, like with engines, the individual modules are responsible for determining if they match that or not.

So on theory we can write a series of PParts specific TF modules that in turn do their own determination and ignore, or supplement, the stated `configuration` property.

**Comment by NathanKell** [comment]
Ok, that makes sense. Because, sure as shootin' I don't think it's a good idea to have dozens, let alone hundreds, of components on a part.

Oh, question: do TF modules that have a configuration mismatch set themselves to inactive? or just don't do anything during their Update loops? Because the former should be (AIUI) much faster.

**Comment by jwvanderbeck** [comment]
Basically the way it works is the TF core interrogates each instance of a loaded module on a part (which is all of them) and asks if it should be active or not.  So each module then decides based on all the data if it is active to the current configuration or not, and reports back. 

So essentially when the API asks for a given module or list of modules on a `Part` the module(s) given back are *only* the ones that would be active for that configuration.

**Comment by NathanKell** [comment]
I'm looking at e.g. https://github.com/KSP-RO/TestFlight/blob/masterTestFlightFailure_IgnitionFail.cs and it has an OnUpdate method. If I'm not mistaken, that means that every single Failure_ignitionFail module added to a part (one per configuration remember, and there can be up to like 10 with engines, and per above hundreds with tanks if we just do cfg-only speccing) will run OnUpdate every frame.

So: does the core actually go ahead and disable the modules that don't have matching configurations? Or do the modules still run their own OnUpdates, and just not do anything if configuration doesn't match?

**Comment by jwvanderbeck** [comment]
In that module, OnUpdate will be called but will immediately drop out if that module doesn't meet the config criteria.

I see what you are saying, and I seem to recall I had issues with the API doing it the way you suggested.

**Comment by NathanKell** [comment]
That's fine for only a few modules, but we're talking like 50 (5 failure modules, 10 conifgurations) for the max case for engines. That's really quite a lot of checks. How hard would it be for a single failure module to support multiple configurations (in the same way that ModuleEngineConfigs passes multiple configs to a single ModuleEngines) ?

**Comment by NathanKell** [comment]
that is, it could have multiple CONFIGURATION nodes, and an internal dict to store them, using the correct one each frame.

**Comment by jwvanderbeck** [comment]
`configuration = $|AJ10` would match every engine with a name that **starts with** AJ10 and thus meeting multiple configs.
You can also do `OR` style configs:
`configuration = AJ10 || RL10`

As long as you want all the various configs to use the same module settings, then you can have one failure module match tons of different parts.  Heck if you use the same failures for every engine in RO, you can use one module for them all.  A n empty config is always a match.

Also bear in mind that not all TF modules have an OnUpdate, though the one you referenced does it isn't universal.  Depends on the module.

**Comment by NathanKell** [comment]
Ya, the problem is that the settings *aren't* the same for different engine configs, that's the point. :D Also...is flight data stored per unique config, or per regexp config? I.e. even if say the AJ10-37 and AJ10-42 had the same curves, we obviously want different data stored for each and different tech transfer of that data.

And yeah, depends on the module, so the hit isn't as bad as it could be.

**Comment by jwvanderbeck** [comment]
> Ya, the problem is that the settings aren't the same for different engine configs

Well I think for many failures they are.  Not ignitions though of course.

> Also...is flight data stored per unique config, or per regexp config? I.e. even if say the AJ10-37 and AJ10-42 had the same curves, we obviously want different data stored for each and different tech transfer of that data.

I need to check the code but I believe the determining factor for recorded data is up to how the TestFlightDataRecorder reports back.  Not the ancillary modules like failures.

**Comment by jwvanderbeck** [comment]
Sorry my brain had what you were asking earlier backwards but I see what you mean now.  You want one module to handle different configs but with different settings for each config.  That is not doable and would definitely require some changes, possible major ones, not sure.  There would still be a check each frame to determine which to use though and I'm not sure the difference in performance would be as big as you think.  Though I could be wrong.

**Comment by NathanKell** [comment]
Yeah, would need to profile to be sure, just my impression is there's a fair amount of overhead Unity adds for each component added to an object. A middle ground might be calling SetActive(false) on the disabled components whenever the configuration of a part changes (i.e. calling that from TestFlightCore). Although, hmm, it's possible Part ignores that flag when running on modules...

**Comment by jwvanderbeck** [comment]
> A middle ground might be calling SetActive(false) on the disabled components whenever the configuration of a part changes (i.e. calling that from TestFlightCore). Although, hmm, it's possible Part ignores that flag when running on modules...

Yeah that is the obvious thing, which is why I am fairly certain I had looked at that in the past and rejected it.  IIRC it just didn't work with the way the TF API was architected.

**Comment by NathanKell** [comment]
Then I'd at least suggest caching the results of EvaluateQuery() and only rerunning it when configuration changes (which the core would be notified about, and could notify all modules). Because right now every module that has an update check, runs that query vs configuration every frame.

**Comment by NathanKell** [comment]
Sorry, caching IsEnabled, which in turn runs EvaluateQuery.

**Comment by jwvanderbeck** [comment]
@NathanKell for these proc parts..would they all basically have (or at least most have) the same exact config just want them to be recognized as different parts, and thus collect data separately?

I think i've worked out a way to do it, just need to clarify the details.
